google/cloud/firestore_v1/query_results.py

```diff
@@ -68,4 +69,4 @@ def get_explain_metrics(self) -> ExplainMetrics:
         if self._explain_options is None:
             raise QueryExplainError("explain_options not set on query.")
         else:
-            return self._explain_metrics
+            return self._explain_metrics  # type: ignore
```

Comment:
> I feel like this is caused by mypy being unaware of the inner logic of the function. It either returns ExplainMetrics, or raises an exception, so there's no way it returns None. I will add # type: ignore to this line.

I don't think that's true. There's nothing stopping someone from passing explain_options, but not explain_metrics:
```
instance = QueryResultList([], object(), None)
instance.get_explain_metrics() == None
```

There are a couple ways we can address this. Maybe we want to ensure that if one is set, they both are, and store them together as a tuple? Or maybe we should just add a separate exception here if self._explain_metrics is None as well? I can't remember off the top of my head if it makes sense to have one but not the other here

---

Either way, I think adding `# type: ignore` should be a last resort. We can sometimes use `typing.cast` if mypy ever has trouble following the invariants. But I think in this case, it was actually a good catch in the logic

Reply:
You are right, mypy did have a good point! When explain_options is set, ultimately server will return explain_metrics, and it should already been gathered before QueryResultList is initialized. To make QueryResultsList more robust, I can do the following:
1. When initializing, throw an error when `explain_options is not None and explain_metrics is None`. (maybe even the case `explain_options is None and explain_metrics is not None`?)
2. Add a check in `get_explain_metrics()` like the following in case `self._explain_metrics` got wiped out somewhere
```
    def get_explain_metrics(self) -> ExplainMetrics:
        if self._explain_options is None:
            raise QueryExplainError("explain_options not set on query.")
        elif self._explain_metrics is None:
            raise QueryExplainError("explain_metrics is empty despite explain_options is set.")
        else:
            return self._explain_metrics
```

Reply:
That works, or you can require them to be passed as a tuple, which would solve the same typing problem with less checks:

```
class QueryResultsList:

    def __init__(
        self,
        docs: List,
        profile_data: tuple[ExplainOptions, ExplainMetrics] | None = None
    ):
```

Reply:
done